src/frequenz/sdk/actor/data_sourcing/microgrid_source.py

```diff
@@ -0,0 +1,374 @@
+}
+
+_InverterDataMethods: Dict[ComponentMetric, Callable[[InverterData], float]] = {
+    ComponentMetric.ACTIVE_POWER: lambda msg: msg.active_power,
+    ComponentMetric.ACTIVE_POWER_LOWER_BOUND: lambda msg: msg.active_power_lower_bound,
+    ComponentMetric.ACTIVE_POWER_UPPER_BOUND: lambda msg: msg.active_power_upper_bound,
+}
+
+_EVChargerDataMethods: Dict[ComponentMetric, Callable[[EVChargerData], float]] = {
+    ComponentMetric.ACTIVE_POWER: lambda msg: msg.active_power,
+    ComponentMetric.CURRENT_PHASE_1: lambda msg: msg.current_per_phase[0],
+    ComponentMetric.CURRENT_PHASE_2: lambda msg: msg.current_per_phase[1],
+    ComponentMetric.CURRENT_PHASE_3: lambda msg: msg.current_per_phase[2],
+    ComponentMetric.VOLTAGE_PHASE_1: lambda msg: msg.voltage_per_phase[0],
+    ComponentMetric.VOLTAGE_PHASE_2: lambda msg: msg.voltage_per_phase[1],
+    ComponentMetric.VOLTAGE_PHASE_3: lambda msg: msg.voltage_per_phase[2],
+}
+
+
+# pylint: disable=fixme
+## FIXME: does `src/frequenz/sdk/data_pipeline/types.py` sounds like a good
+## place for the `Sample` class?
+@dataclass(frozen=True)
+class Sample:
+    """Describes a single observed floating point value at a specific timestamp."""
+
+    timestamp: datetime
+    value: Optional[float] = None
+
+
+class MicrogridApiSource:
+    """Fetches requested metrics from the Microgrid API.
+
+    Used by the DataSourcingActor.
+    """
+
+    def __init__(
+        self,
+        registry: ChannelRegistry,
+    ) -> None:
+        """Create a `_MicrogridApiSource` instance.
+
+        Args:
+            registry: A channel registry.  To be replaced by a singleton
+                instance.
+        """
+        self.comp_data_receivers: Dict[int, Receiver[Any]] = {}
+        self.comp_data_tasks: Dict[int, asyncio.Task[None]] = {}
+        self._registry = registry
+        self._req_streaming_metrics: Dict[
+            int, Dict[ComponentMetric, List[ComponentMetricRequest]]
+        ] = {}
+
+    async def _check_battery_request(
+        self,
+        comp_id: int,
+        requests: Dict[ComponentMetric, List[ComponentMetricRequest]],
+    ) -> None:
+        """Check if the requests are valid Battery metrics.
+
+        Raises:
+            TypeError: if the requested metric is not available for batteries.
+
+        Args:
+            comp_id: The id of the requested component.
+            requests: A list of metric requests received from external actors
+                for the given battery.
+        """
+        for metric in requests.keys():
+            if metric not in {
+                ComponentMetric.SOC,
+                ComponentMetric.SOC_LOWER_BOUND,
+                ComponentMetric.SOC_UPPER_BOUND,
+                ComponentMetric.CAPACITY,
+                ComponentMetric.POWER_LOWER_BOUND,
+                ComponentMetric.POWER_UPPER_BOUND,
+            }:
+                raise TypeError(f"Unknown metric {metric} for Battery id {comp_id}")
```

Comment:
Nitpick: I would use a `ValueError` here, as `TypeError` are supposed to be mismatches on the type system.

Reply:
done